// saucelabs/build/sauceConnectLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = require("fs");

var _util = require("util");

var _path = require("path");

var _download = _interopRequireDefault(require("download"));

var _constants = require("./constants");

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Note:
 * This is a modified fork of https://github.com/kevva/bin-wrapper
 * The module had a CVE-2021-3795 vulnerability and decided to fork
 * and "simplify" the implementation
 *
 * Original licence:
 *
 * MIT License
 * Copyright (c) Kevin Mårtensson <kevinmartensson@gmail.com>
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
 * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */
class SauceConnectLoader {
  constructor(options = {}) {
    const platform = (0, _utils.getPlatform)();
    const platformData = _constants.SAUCE_CONNECT_PLATFORM_DATA[platform];

    if (!platformData) {
      throw new ReferenceError(`Unsupported platform ${platform}`);
    }

    const {
      url,
      use
    } = platformData;
    this.url = (0, _util.format)(url, options.sauceConnectVersion);
    this.dest = (0, _path.join)(__dirname, 'sc-loader', `.sc-v${options.sauceConnectVersion}`);
    this.path = (0, _path.join)(this.dest, use);
  }
  /**
   * Verify if SC was already downloaded,
   * if not then download it
   *
   * @api public
   */


  verifyAlreadyDownloaded() {
    return _fs.promises.stat(this.path).catch(err => {
      if ((err === null || err === void 0 ? void 0 : err.code) === 'ENOENT') {
        return this._download();
      }

      throw err;
    });
  }
  /**
   * Download SC
   *
   * @api private
   */


  _download() {
    return (0, _download.default)(this.url, this.dest, {
      extract: true,
      strip: 1
    }).then(() => {
      if (process.platform !== 'win32') {
        // ensure the sc executable is actually executable
        return _fs.promises.chmod(this.path, 0o755);
      }
    });
  }

}

exports.default = SauceConnectLoader;
